src/evqueue_wfmanager.8
.TH "EVQUEUE_WFMANAGER" 8 "2016-07-11" "evQueue utilities" "evqueue"

.SH NAME
evqueue_wfmanager \- evQueue utilities
.SH "SYNOPSIS"
.PP
\fBevqueue_wfmanager\fR [ OPTIONS ] --launch <wf name> [--<parameter name> <parameter value>]*
.SH "DESCRIPTION"
.PP
Utility to launch workflows from command line. Can be used to connect to the evQueue core through TCP or UNIX sockets.
.PP
Options MUST be passed before --launch argument. Arguments expected after are only workflow parameters.
.SH "OPTIONS"
.TP
\fB--connect\fR \fIconnection_string\fR
Connect to server identified by \fIconnection_string\fR. Two formats are accepted :

tcp socket: tcp://<hostname>[:<port>]

unix socket: unix://<path>
.TP
\fB--user <username>\fR
Use this username if authentication is required by daemon.
.TP
\fB--password <password>\fR
Use this password if authentication is required by daemon.
.TP
\fB--mode [synchronous|asynchronous]\fR
Run in synchronous or asynchronous mode. In synchronous mode, the command will only return after the workflow is terminated. In asynchronous mode, the command returns immediately after the workflow is launched.
.TP
\fB--timeout <timeout>\fR
When running in synchronous mode, do not wait worfklow termination for more than <timeout> seconds.
.SH "CONFIGURATION FILES"
.PP
evqueue_wfmanager will read configuration files in /etc/evqueue.api.conf and ~/.evqueue.api.conf
.PP
The following options are recognized (but will be overridden by command line parameters) : api.connect, api.user, api.password
.SH EXAMPLES
evqueue_wfmanager --connect tcp://localhost:5000 --user admin --password admin --launch sleep --time 1
.br
Connects to evQueue daemon using TCP on localhost (port 5000). Launches a workflow named "sleep" which takes on parameter "time" (whose value is "1").
.SH SEE ALSO
.BR http://www.evqueue.net/



debian/changelog
evqueue-core (3.2-1) buster; urgency=medium

  * Automatic tags
  * Bugfixes

 -- Thibault KUMMER <bob@coldsource.net>  Mon, 14 Dec 2020 10:54:16 +0200

evqueue-core (3.1-1) buster; urgency=medium

  * Bugfixes

 -- Thibault KUMMER <bob@coldsource.net>  Tue, 10 Nov 2020 12:30:16 +0200

evqueue-core (3.0-1) buster; urgency=medium

  * Websockets support

 -- Thibault KUMMER <bob@coldsource.net>  Thu, 14 May 2020 12:30:16 +0200

evqueue-core (2.1-1) buster; urgency=medium

  * Fix: run subjobs when loop has no iterations
  * Fix: segfault
  * Fix: restart rsyslog only if present (closes: #881596)
  * Added debconf translations (closes: #882654, #883290, #883302, #883822,
    #884171)

 -- Thibault KUMMER <bob@coldsource.net>  Thu, 08 Nov 2018 09:52:40 +0000

evqueue-core (2.0-2) buster; urgency=medium

  * Initial release (closes: #880973).

 -- Thibault KUMMER <bob@coldsource.net>  Wed, 17 Aug 2016 12:30:16 +0200


src/evqueue_wfmanager.cpp
/*
 * This file is part of evQueue
 * 
 * evQueue is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * evQueue is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with evQueue. If not, see <http://www.gnu.org/licenses/>.
 * 
 * Author: Thibault Kummer <bob@coldsource.net>
 */

#include <API/ClientBase.h>
#include <API/XMLResponse.h>
#include <Exception/Exception.h>
#include <XML/XMLString.h>
#include <Configuration/ConfigurationReader.h>
#include <Configuration/Configuration.h>

#include <stdio.h>
#include <string.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <sys/un.h>
#include <netdb.h>

#include <map>
#include <string>
#include <iomanip>
#include <sstream>

#include <xercesc/dom/DOM.hpp>

using namespace std;

static void usage()
{
	fprintf(stderr,"Usage : evqueue_wfmanager --launch <wf name> --<parameter name> <parameter value>\n");
	fprintf(stderr,"  --connect tcp://<hostname>[:<port>] | unix://<path>\n");
	fprintf(stderr,"  --user <username>\n");
	fprintf(stderr,"  --password <password>\n");
	fprintf(stderr,"  --mode [synchronous|asynchronous]\n");
	fprintf(stderr,"  --timeout <timeout>\n");
	fprintf(stderr,"  --comment <comment>\n");
	exit(-1);
}

int main(int argc, char  **argv)
{
	int exit_status = 0;
	
	try
	{
		// Default config
		Configuration config({{"api.connect","tcp://localhost:5000"}, {"api.user", ""}, {"api.password", ""}});
		
		// Try to read from default paths
		ConfigurationReader::ReadDefaultPaths("evqueue.api.conf", &config);
		
		// Override with command line
		int cmd_cur = ConfigurationReader::ReadCommandLine(argc, argv, {"connect", "user", "password", "bool:noformat"}, "api", &config);
		if(cmd_cur==-1)
			usage();

		string connection_str = config.Get("api.connect");
		string user = config.Get("api.user");
		string password = config.Get("api.password");
		
		// Parse cmdline parameters
		char *workflow_name = 0;
		char *workflow_mode = 0;
		char *workflow_comment = 0;
		char *timeout = 0;
		
		int cur;
		for(cur=cmd_cur;cur<argc;cur++)
		{
			if(strcmp(argv[cur],"--launch")==0)
			{
				if(cur+1>=argc)
				{
					fprintf(stderr,"Missing workflow name : --launch <workflow name>\n");
					return -1;
				}
				
				workflow_name = argv[cur+1];
				cur+=2;
				break;
			}
			else if(strcmp(argv[cur],"--mode")==0)
			{
				if(cur+1>=argc)
				{
					fprintf(stderr,"Missing mode : --mode [synchronous|asynchronous]\n");
					return -1;
				}
				
				workflow_mode = argv[cur+1];
				cur++;
			}
			else if(strcmp(argv[cur],"--timeout")==0)
			{
				if(cur+1>=argc)
				{
					fprintf(stderr,"Missing timeout : --timeout <timeout>\n");
					return -1;
				}
				
				timeout = argv[cur+1];
				cur++;
			}
			else if(strcmp(argv[cur],"--comment")==0)
			{
				if(cur+1>=argc)
				{
					fprintf(stderr,"Missing comment : --mode <comment>\n");
					return -1;
				}
				
				workflow_comment = argv[cur+1];
				cur++;
			}
			else
			{
				fprintf(stderr,"Unknown parameter name : %s\n",argv[cur]);
				return -1;
			}
		}
		
		if(!workflow_name)
			usage();
		
		// If we received a password, compute sha1 of it as a real password
		if(password.length())
			password = ClientBase::HashPassword(password);
		
		// Build workflow XML
		xercesc::XMLPlatformUtils::Initialize();
		
		{
			DOMDocument xmldoc;
			DOMElement workflow_node = xmldoc.createElement("instance");
			xmldoc.appendChild(workflow_node);
			
			workflow_node.setAttribute("action","launch");
			workflow_node.setAttribute("name",workflow_name);
			if(workflow_mode)
				workflow_node.setAttribute("mode",workflow_mode);
			if(timeout)
				workflow_node.setAttribute("timeout",timeout);
			if(workflow_comment)
				workflow_node.setAttribute("comment",workflow_comment);
			
			int status = 1;
			char *parameter_name, *parameter_value;
			for(int i=cur;i<argc;i++)
			{
				if(status==1)
				{
					// Parameter name expected
					if(strncmp(argv[i],"--",2)!=0 || strlen(argv[i])<=2)
					{
						fprintf(stderr,"Parameter name expected with format : --<parameter name>\n");
						return -1;
					}
					
					parameter_name = argv[i]+2;
					status = 2;
				}
				else if(status==2)
				{
					// Value expacted
					parameter_value = argv[i];
					status=1;
					
					DOMElement parameter_node = xmldoc.createElement("parameter");
					parameter_node.setAttribute("name",parameter_name);
					parameter_node.setAttribute("value",parameter_value);
					workflow_node.appendChild(parameter_node);
				}
			}
			
			string workflow_xml = xmldoc.Serialize(workflow_node);
			
			// Send launch command to evQueue
			
			try
			{
				ClientBase client(connection_str,user,password);
				client.Exec(workflow_xml);
				
				XMLResponse *response = client.GetResponseHandler();
				string wfid = response->GetRootAttribute("workflow-instance-id");
				printf("Launched instance %s\n",wfid.c_str());
				
			}
			catch(Exception &e)
			{
				fprintf(stderr,"Exception in client : %s\n",e.error.c_str());
				exit_status = -1;
			}
		}
	}
	catch(Exception &e)
	{
		fprintf(stderr,"%s",e.error.c_str());
		if(e.code!="")
			fprintf(stderr," (%s)",e.code.c_str());
		fprintf(stderr,"\n");
		exit_status = -1;
	}
	
	xercesc::XMLPlatformUtils::Terminate();
	
	return exit_status;
}
